sys/net/gnrc/application_layer/tftp/gnrc_tftp.c

```diff
@@ -0,0 +1,749 @@
+
+#include "net/gnrc/tftp.h"
+#include "net/gnrc/netapi.h"
+#include "net/gnrc/netreg.h"
+#include "net/gnrc/udp.h"
+#include "net/gnrc/ipv6.h"
+#include "random.h"
+
+#define ENABLE_DEBUG                (1)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+/* For PRIu16 etc. */
+#include <inttypes.h>
+#endif
+
+#define ENABLE_TFTP_ERROR_SEND                 0
+
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#define CT_HTONS(x)         ((          \
+                   (x >> 8) & 0x00FF) | \
+                   ((x << 8) & 0xFF00))
+
+#define CT_HTONL(x)         ((          \
+              (x >> 24) & 0x000000FF) | \
+              ((x >> 8) & 0x0000FF00) | \
+              ((x << 8) & 0x00FF0000) | \
+              ((x << 24) & 0xFF000000))
+#else
+#define CT_HTONS(x)    (x)
+#define CT_HTONL(x)    (x)
+#endif
+
+#define MIN(a,b)            ((a) > (b) ? (b) : (a))
+#define ARRAY_LEN(x)        (sizeof(x) / sizeof(x[0]))
+
+/**
+ * @brief TFTP opcodes
+ */
+typedef enum {
+    TO_RRQ          = CT_HTONS(1),      /**< Read Request */
+    TO_RWQ          = CT_HTONS(2),      /**< Write Request */
+    TO_DATA         = CT_HTONS(3),      /**< Data */
+    TO_ACK          = CT_HTONS(4),      /**< Acknowledgment */
+    TO_ERROR        = CT_HTONS(5),      /**< Error */
+    TO_OACK         = CT_HTONS(6),      /**< Option ACK */
+} tftp_opcodes_t;
+
+/**
+ * @brief TFTP Error Codes
+ */
+typedef enum {
+    TE_UN_DEF       = CT_HTONS(0),      /**< Not defined, see error message */
+    TE_NO_FILE      = CT_HTONS(1),      /**< File not found */
+    TE_ACCESS       = CT_HTONS(2),      /**< Access violation */
+    TE_DFULL        = CT_HTONS(3),      /**< Disk full or allocation exceeded */
+    TE_ILLOPT       = CT_HTONS(4),      /**< Illegal TFTP operation */
+    TE_UNKOWN_ID    = CT_HTONS(5),      /**< Unknown transfer ID */
+    TE_EXISTS       = CT_HTONS(6),      /**< File already exists */
+    TE_UNKOWN_USR   = CT_HTONS(7),      /**< No such user */
+} tftp_err_codes_t;
+
+/**
+ * @brief TFTP Transfer modes
+ */
+typedef enum {
+    ASCII,
+    OCTET,
+    MAIL
+} tftp_mode_t;
+
+/**
+ * @brief TFTP Options
+ */
+typedef enum {
+    TOPT_BLKSIZE,
+    TOPT_TIMEOUT,
+    TOPT_TSIZE,
+} tftp_options_t;
+
+typedef enum {
+    FAILED      = -1,
+    BUSY        = 0,
+    FINISHED    = 1
+} tftp_state;
+
+/**
+ * @brief The TFTP context for the current transfer
+ */
+typedef struct {
+    const char *file_name; /* TODO [GNRC_TFTP_MAX_FILENAME_LEN]; */
+    tftp_mode_t mode;
+    tftp_opcodes_t op;
+    ipv6_addr_t *peer;
+    vtimer_t timer;
+    uint32_t timeout;
+    uint16_t dst_port;
+    uint16_t src_port;
+    tftp_data_callback cb;
+
+    /* transfer parameters */
+    uint16_t block_nr;
+    uint16_t block_size;
+    uint32_t transfer_size;
+    uint16_t block_timeout;
+    bool use_options;
+} tftp_context_t;
+
+/**
+ * @brief The default TFTP header
+ */
+typedef struct __attribute__((packed)) {
+    tftp_opcodes_t opc          : 16;
```

Comment:
Because there are no "compile time" HTONS NTONS macro's. Which means that I couldn't define my enums like:

	typedef enum {
		TO_RRQ          = CT_HTONS(1),      /**< Read Request */
		TO_RWQ          = CT_HTONS(2),      /**< Write Request */
		TO_DATA         = CT_HTONS(3),      /**< Data */
		TO_ACK          = CT_HTONS(4),      /**< Acknowledgment */
		TO_ERROR        = CT_HTONS(5),      /**< Error */
		TO_OACK         = CT_HTONS(6),      /**< Option ACK */
	} tftp_opcodes_t;

And don't have to swap the error and op codes runtime time. I could change them to network_uint16_t's and use defines, but this is easier for debugging :)